src/solvers/twostep_time_solver.C

```diff
@@ -257,4 +287,103 @@ void TwostepTimeSolver::solve()
     }
 }
 
+std::pair<unsigned int, Real> TwostepTimeSolver::adjoint_solve (const QoISet & qoi_indices)
+{
+  // The adjoint timestepping mirrors the scheme used for the forward problem
+  // So the deltat, once set by solution history, will not be changed
+  Real old_time = _system.time;
+
+  // Take the first adjoint 'half timestep'
+  core_time_solver->adjoint_solve(qoi_indices);
+
+  // Adjoint advance the timestep
+  core_time_solver->adjoint_advance_timestep();
+
+  // The second half timestep
+  std::pair<unsigned int, Real> full_adjoint_output = core_time_solver->adjoint_solve(qoi_indices);
+
+  // Tell our timesolver the combined deltat before adjoint advance timestep updates it
+  this->completedtimestep_deltat = 2.0*_system.deltat;
+
+  // Adjoint advance again so solution history saves the adjoint
+  core_time_solver->adjoint_advance_timestep();
+
+  // Reset the system.time
+  _system.time = old_time;
+
+  return full_adjoint_output;
+}
+
+void TwostepTimeSolver::integrate_adjoint_sensitivity(const QoISet & qois, const ParameterVector & parameter_vector, SensitivityData & sensitivities)
+{
+  // We are using the midpoint rule to integrate each timestep
+  // (f(t_j) + f(t_j+1/2))/2 (t_j+1/2 - t_j) + (f(t_j+1/2) + f(t_j+1))/2 (t_j+1 - t_j+1/2)
+
+  // Get t_j
+  Real time_left = _system.time;
+
+  // Left side sensitivities to hold f(t_j)
+  SensitivityData sensitivities_left(qois, _system, parameter_vector);
+
+  // Get f(t_j)
+  _system.adjoint_qoi_parameter_sensitivity(qois, parameter_vector, sensitivities_left);
+
+  // Advance to t_j+1/2
+  _system.time = _system.time + _system.deltat;
+
+  // Get t_j+1/2
+  Real time_middle = _system.time;
+
+  // Middle sensitivities f(t_j+1/2)
+  SensitivityData sensitivities_middle(qois, _system, parameter_vector);
+
+  // Remove the sensitivity rhs vector from system since we did not write it to file and it cannot be retrieved
+  _system.remove_vector("sensitivity_rhs0");
```

Comment:
This is going to be a bunch of deallocation and reallocation we could skip - I thought we'd decided to have a read option that lets us ignore vectors that are in the system but not in the save file.

Reply:
We discussed over chat that the changes to equationsystems::read should be in a separate PR.

Reply:
Right - but when you asked me that I'd assumed you meant an *earlier* PR, not a later one.  It ought to be both much simpler than and a natural dependency of this one.